File: C++/RecoverBinarySearchTree.cc
// Two elements of a binary search tree (BST) are swapped by mistake.

// Recover the tree without changing its structure.

class Solution {
    TreeNode* prev{nullptr};
    TreeNode* wrong{nullptr};
    TreeNode* cand{nullptr};

    void recover(TreeNode* root) {
        if (!root) {
            return;
        }
        recover(root->left);
        if (prev && prev->val > root->val) {
            if (!wrong) {
                wrong = prev;
                cand = root;
            } else {
                swap(root->val, wrong->val);
                wrong = nullptr;
                return;
            }
        }
        prev = root;
        recover(root->right);
    }
public:
    void recoverTree(TreeNode* root) {
        recover(root);
        if (wrong) {
            swap(cand->val, wrong->val);
        }
    }
};

File: Java/RecoverBinarySearchTree.java
// Two elements of a binary search tree (BST) are swapped by mistake.

// Recover the tree without changing its structure.

// Note:
// A solution using O(n) space is pretty straight forward. Could you devise a constant space solution?

public class RecoverBinarySearchTree {
    TreeNode prev = null;
    TreeNode n1 = null;
    TreeNode n2 = null;

    public void recoverTree(TreeNode root) {
        findWrongNode(root);
        if (n1 != null || n2 != null) {
            int temp = n1.val;
            n1.val = n2.val;
            n2.val = temp;
        }
    }

    void findWrongNode(TreeNode node) {
        if (node == null) {
            return;
        }

        findWrongNode(node.left);
        
        if (prev != null && prev.val > node.val){
            if (n1 == null) {
                n1 = prev;
            }
            n2 = node;
        }
        prev = node;
        
        findWrongNode(node.right);
    }
}
